fl4health/strategies/feddg_ga_strategy.py

```diff
@@ -0,0 +1,411 @@
+        weight_step_size: float = 0.2,
+    ):
+        """Strategy for the FedDG-GA algorithm (Federated Domain Generalization with
+        Generalization Adjustment, Zhang et al. 2023).
+
+        Args:
+            fraction_fit : float, optional
+                Fraction of clients used during training. In case `min_fit_clients`
+                is larger than `fraction_fit * available_clients`, `min_fit_clients`
+                will still be sampled. Defaults to 1.0.
+            fraction_evaluate : float, optional
+                Fraction of clients used during validation. In case `min_evaluate_clients`
+                is larger than `fraction_evaluate * available_clients`, `min_evaluate_clients`
+                will still be sampled. Defaults to 1.0.
+            min_fit_clients : int, optional
+                Minimum number of clients used during training. Defaults to 2.
+            min_evaluate_clients : int, optional
+                Minimum number of clients used during validation. Defaults to 2.
+            min_available_clients : int, optional
+                Minimum number of total clients in the system. Defaults to 2.
+            evaluate_fn :
+                Optional[
+                    Callable[[int, NDArrays, Dict[str, Scalar]],
+                    Optional[Tuple[float, Dict[str, Scalar]]]]
+                ]
+                Optional function used for validation. Defaults to None.
+            on_fit_config_fn : Callable[[int], Dict[str, Scalar]], optional
+                Function used to configure training. Defaults to None.
+            on_evaluate_config_fn : Callable[[int], Dict[str, Scalar]], optional
+                Function used to configure validation. Defaults to None.
+            accept_failures : bool, optional
+                Whether or not accept rounds containing failures. Defaults to True.
+            initial_parameters : Parameters, optional
+                Initial global model parameters.
+            fit_metrics_aggregation_fn : Optional[MetricsAggregationFn]
+                Metrics aggregation function, optional.
+            evaluate_metrics_aggregation_fn : Optional[MetricsAggregationFn]
+                Metrics aggregation function, optional.
+            fairness_metric : FairnessMetric, optional.
+                The metric to evaluate the local model of each client against the global model in order to
+                determine their adjustment weight for aggregation. Can be set to any default metric in
+                FairnessMetricType or set to use a custom metric. Optional, default is
+                FairnessMetric(FairnessMetricType.LOSS).
+            weight_step_size : float
+                The step size to determine the magnitude of change for the adjustment weight. It has to be
+                0 < weight_step_size < 1. Optional, default is 0.2.
+        """
+        super().__init__(
+            fraction_fit=fraction_fit,
```

Comment:
For this method, one thing I've been thinking about is that it's sort of "ill-defined" for situations where you only have some of the clients participating in any given round. The paper doesn't really contemplate this anywhere that I can find. Because of that, I think it would be good practice to warn the user that the algorithms convergence and behavior are not guaranteed if `fraction_fit` and `fraction_evaluate` are not 1.0, even though it is technically supported? No big changes needed, just logging a warning is sufficient.